internal/controllers/topology/cluster/reconcile_state.go

```diff
@@ -307,6 +307,18 @@ func (r *Reconciler) reconcileInfrastructureCluster(ctx context.Context, s *scop
 // reconcileControlPlane works to bring the current state of a managed topology in line with the desired state. This involves
 // updating the cluster where needed.
 func (r *Reconciler) reconcileControlPlane(ctx context.Context, s *scope.Scope) error {
+	// If the ControlPlane has defined a current or desired MachineHealthCheck attempt to reconcile it.
+	if s.Desired.ControlPlane.MachineHealthCheck != nil || s.Current.ControlPlane.MachineHealthCheck != nil {
+		// Reconcile the current and desired state of the MachineHealthCheck.
+		if err := r.reconcileMachineHealthCheck(ctx, s.Current.ControlPlane.MachineHealthCheck, s.Desired.ControlPlane.MachineHealthCheck); err != nil {
+			return err
+		}
+	}
```

Comment:
Refactored to not block MHC reconciliation if the control plane is pending an upgrade.

(Same for MD)